File: src/controls/text_field.ts
import { observable } from "../core/decorators";
import { Control } from "./control";

// NSTextField analog. Wraps an `<input>` (or an element containing one).
// Two-way: `input` event → `value`; assigning `value` writes the DOM.
// Subclass with `@register` to expose under `klass="…"`.
export class TextField extends Control {
  @observable accessor value: string = "";

  constructor(el: HTMLElement) {
    super(el);
    const input = this.inputElement;
    if (input) {
      if (input.value !== this.value) this.value = input.value;
      input.addEventListener("input", this.onInput);
    }
  }

  protected get inputElement(): HTMLInputElement | null {
    const el = this.element;
    if (el instanceof HTMLInputElement) return el;
    return el.querySelector("input");
  }

  private onInput = (e: Event): void => {
    const t = e.target as HTMLInputElement;
    if (t === this.inputElement && t.value !== this.value) this.value = t.value;
  };

  valueDidChange(_prev: string, next: string): void {
    const input = this.inputElement;
    if (input && input.value !== next) input.value = next;
  }
}

File: opal/lib/swill/text_field.rb
# backtick_javascript: true

module Swill
  # NSTextField-style observable wrapper for an input or input-containing view.
  class TextField < Control
    include Observable

    property :value, default: ""

    def initialize(element)
      super
      input = input_element
      if input
        initial = `#{input}.value`
        initial = `#{input}.getAttribute("value")` if (initial.nil? || initial == "") && `#{input}.hasAttribute("value")`
        self.value = initial.to_s
      end
      @input_listener = lambda do |event|
        target = `#{event}.target`
        self.value = `#{target}.value`.to_s if `#{target} === #{input_element}`
      end
      `#{input}.addEventListener("input", #{@input_listener})` if input
    end

    def value_did_change(_previous, value)
      input = input_element
      `#{input}.value = #{value.to_s}` if input && `#{input}.value !== #{value.to_s}`
    end

    def teardown!
      input = input_element
      `#{input}.removeEventListener("input", #{@input_listener})` if input && @input_listener
    end

    private

    def input_element
      `#{element}.tagName === "INPUT" ? #{element} : #{element}.querySelector("input")`
    end
  end
end
